### PR TITLE
rgw: fix Metadata search is not available when using tenants

### PR DESCRIPTION
**Pull request by xxcs**
http://tracker.ceph.com/issues/39940

Fixes: http://tracker.ceph.com/issues/39940
Signed-off-by: Snow Si  silonghu@inspur.com

src/rgw/rgw_sync_module_es_rest.cc: modify permissions ->permissions.keyword

rgw: fix Metadata search is not available when using tenants
tenant$uid: xxcs$test
permissions: xxcs$test
a.Elasticsearch PUT OBJECT: (permissions: xxcs)->OBJECT; (permissions: test)->OBJECT (permissions.keyword: xxcs$test) -> OBJECT
b.Elasticsearch Query: (permissions == xxcs$test) can not -> OBJECT; (permissions.keyword == xxcs$test) can -> OBJECT

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by xxcs** [comment]
Signed-off-by: Snow Si 2804536059@qq.com

**Comment by smithfarm** [comment]
@xxcs The following two lines should be added to the commit message itself (they can be in the PR description, too, but the git history is decisive):

```
Fixes: http://tracker.ceph.com/issues/39940
Signed-off-by: Snow Si <2804536059@qq.com>
```

**Comment by smithfarm** [comment]
By the way, you can use `git log` to peruse the git history of the master branch - many examples of commit messages there.

**Comment by xxcs** [comment]
OK, I have just modified it according to your suggestion, Thank you for your comments

**Comment by xxcs** [comment]
> By the way, you can use `git log` to peruse the git history of the master branch - many examples of commit messages there.

OK, I have just modified it according to your suggestion, Thank you for your comments

**Comment by xxcs** [comment]
> http://tracker.ceph.com/issues/39940
> 
> Fixes: http://tracker.ceph.com/issues/39940
> Signed-off-by: Snow Si [silonghu@inspur.com](mailto:silonghu@inspur.com)
> 
> * [x]  References tracker ticket
> * [x]  Updates documentation if necessary
> * [x]  Includes tests for new functionality or reproducer for bug

**Comment by smithfarm** [comment]
@xxcs The title (first line) of the commit message should be prefixed with the component the commit is touching, and should briefly describe what the change is doing. So, in this case, please change it to something like:

`rgw: fix Metadata search is not available with certain tenants`

**Comment by xxcs** [comment]
> @xxcs The title (first line) of the commit message should be prefixed with the component the commit is touching, and should briefly describe what the change is doing. So, in this case, please change it to something like:
> 
> `rgw: fix Metadata search is not available with certain tenants`

Ok, because "make check" can not pass, So I am trying to modify it.

**Comment by smithfarm** [comment]
@xxcs Changing the commit message should be a simple matter of `git commit --amend` and then `git push -f`

And I see that "make check" has passed now ;-)

**Comment by xxcs** [comment]
> @xxcs Changing the commit message should be a simple matter of `git commit --amend` and then `git push -f`
> 
> And I see that "make check" has passed now ;-)

Hello, are there any other questions?

**Comment by xxcs** [comment]
@yehudasa @theanalyst 
    Hello, Can you help review the code? Thank you very much.

File: src/rgw/rgw_sync_module_es_rest.cc [hunk 1]
```diff
@@ -204,7 +207,7 @@ void RGWMetadataSearchOp::execute()
   ESEntityTypeMap gm(generic_map);
   es_query.set_generic_type_map(&gm);
 
-  static set<string> restricted_fields = { {"permissions"} };
+  // static set<string> restricted_fields = { {"permissions.keyword"} };
```
Comment (yuvalif) [inline src/rgw/rgw_sync_module_es_rest.cc]: please remove dead code

File: src/rgw/rgw_es_query.cc [hunk 2]
```diff
@@ -149,15 +149,20 @@ class ESQueryNode_Bool : public ESQueryNode {
   }
   virtual ~ESQueryNode_Bool() {
     delete first;
-    delete second;
+    if (second) {
```
Comment (yuvalif) [inline src/rgw/rgw_es_query.cc]: not needed. `delete nullptr` is just no-op

File: src/rgw/rgw_sync_module_es_rest.cc (hunk 1 above)
Comment (xxcs) [reply src/rgw/rgw_sync_module_es_rest.cc]: ok

File: src/rgw/rgw_es_query.cc (hunk 2 above)
Comment (xxcs) [reply src/rgw/rgw_es_query.cc]: Yes, you are right

File: src/rgw/rgw_es_query.cc [hunk 3]
```diff
@@ -149,15 +149,19 @@ class ESQueryNode_Bool : public ESQueryNode {
   }
   virtual ~ESQueryNode_Bool() {
     delete first;
+    first = nullptr;
     delete second;
+    second = nullptr;
```
Comment (theanalyst) [inline src/rgw/rgw_es_query.cc]: this as well

**Comment by theanalyst** [comment]
can you see if this works with the latest master, since we changed str in ES versions >= 5 since it's deprecated

**Comment by xxcs** [comment]
> can you see if this works with the latest master, since we changed str in ES versions >= 5 since it's deprecated

The Elasticsearch'version which I have been using is 5.6.9

**Comment by theanalyst** [comment]
> > can you see if this works with the latest master, since we changed str in ES versions >= 5 since it's deprecated
> 
> The Elasticsearch'version which I have been using is 5.6.9

understood, but the str/keyword fix is already in ceph/master, have you tested this against ceph/master?

**Comment by theanalyst** [comment]
this is the pr https://github.com/ceph/ceph/pull/26106 (which is in nautilus) which changes string to text for ES >= 5.0 so that should likely help the issue you're seeing

**Comment by theanalyst** [comment]
this is the pr https://github.com/ceph/ceph/pull/26106 (which is in nautilus) which changes string to text for ES >= 5.0 so that should likely help the issue you're seeing

**Comment by xxcs** [comment]
> this is the pr #26106 (which is in nautilus) changes string to text for ES >= 5.0 so that should likely help the issue you're seeing

The results of my test have been based on ceph/master, as follows("permissions" in mapping of the index):
			"permissions": {
				"type": "text",
				"fields": {
					"keyword": {
						"type": "keyword",
						"ignore_above": 256
					}
				}
			}

**Comment by theanalyst** [comment]
unstale me

**Pull request closed without merging**